src/AspNet.Security.OAuth.Bilibili/BilibiliAuthenticationHandler.cs

```diff
@@ -0,0 +1,220 @@
+        };
+
+        foreach (var additionalParameter in Options.AdditionalAuthorizationParameters)
+        {
+            parameters.Add(additionalParameter.Key, additionalParameter.Value);
+        }
+
+        parameters["state"] = Options.StateDataFormat.Protect(properties);
+
+        return QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, parameters);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+    {
+        // See https://open.bilibili.com/doc/4/eaf0e2b5-bde9-b9a0-9be1-019bb455701c#h1-u7B80u4ECB for details.
+        var tokenRequestParameters = new Dictionary<string, string?>()
+        {
+            ["client_id"] = Options.ClientId,
+            ["code"] = context.Code,
+            ["client_secret"] = Options.ClientSecret,
+            ["grant_type"] = "authorization_code",
+        };
+
+        using var tokenRequestContent = new FormUrlEncodedContent(tokenRequestParameters);
+
+        using var response = await Backchannel.PostAsync(Options.TokenEndpoint, tokenRequestContent, Context.RequestAborted);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.AccessTokenError(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+        }
+
+        using var stream = await response.Content.ReadAsStreamAsync(Context.RequestAborted);
+        using var document = await JsonDocument.ParseAsync(stream);
+
+        var mainElement = document.RootElement;
+        if (!ValidateReturnCode(mainElement, out var code))
+        {
+            return OAuthTokenResponse.Failed(new Exception($"An error (Code:{code}) occurred while retrieving an access token."));
+        }
+
+        var payload = JsonDocument.Parse(mainElement.GetProperty("data").GetRawText());
+        return OAuthTokenResponse.Success(payload);
+    }
+
+    private static string ComputeHmacSHA256(string key, string data)
+    {
+        var keyBytes = Encoding.UTF8.GetBytes(key);
+        var dataBytes = Encoding.UTF8.GetBytes(data);
+        var hash = HMACSHA256.HashData(keyBytes, dataBytes);
+        return Convert.ToHexStringLower(hash);
+    }
+
+#pragma warning disable CA5351
+    private static string ComputeMd5(string input)
+    {
+        var inputBytes = Encoding.ASCII.GetBytes(input);
+        var hashBytes = MD5.HashData(inputBytes);
+        return Convert.ToHexStringLower(hashBytes);
+    }
+#pragma warning disable CA5351
+
+    private static string BuildSignatureString(HttpRequestMessage request, string appSecret)
+    {
+        var headers = request.Headers
+            .Where(h => h.Key.StartsWith("x-bili-", StringComparison.OrdinalIgnoreCase))
+            .OrderBy(h => h.Key)
+            .Select(h => $"{h.Key}:{string.Join(",", h.Value)}")
+            .ToList();
+
+        var signature = string.Join("\n", headers);
```

Comment:
Small optimisation to use the char overload:

```suggestion
        var signature = string.Join('\n', headers);
```